miniapps/tools/least-squares.cpp

```diff
@@ -0,0 +1,1720 @@
+}
+
+/// @brief Dense small least squares solver, with constrains @a C with value @a c
+void LSSolver(Solver& solver,
+              const DenseMatrix& A,
+              const DenseMatrix& C,
+              const Vector& b,
+              const Vector& c,
+              Vector& x,
+              Vector& y,
+              real_t shift = 0.0,
+              DenseMatrix* M = nullptr)
+{
+   Array<int> offsets(3);
+   offsets[0] = 0;
+   offsets[1] = A.Width();
+   offsets[2] = A.Width() + C.Height();
+
+   // Block vectors
+   BlockVector rhs(offsets), z(offsets);
+   rhs = 0.0;
+   z = 0.0;
+
+   Vector Atb(A.Width());
+   A.MultTranspose(b, Atb);
+
+   rhs.SetVector(Atb, offsets[0]);
+   rhs.SetVector(c, offsets[1]);
+
+   auto AtA_reg = OperatorPtr(Operator::Type::ANY_TYPE);
+   auto _M = OperatorPtr(Operator::Type::ANY_TYPE);
+   auto block_solver = OperatorPtr(Operator::Type::ANY_TYPE);
+
+   if (auto it_solver = dynamic_cast<IterativeSolver*>(&solver))
+   {
+      if (M) { _M.Reset(M, false); }
+      else { _M.Reset(new IdentityOperator(A.Width())); }
+
+      auto _At = new TransposeOperator(A);
+      auto _AtA = new ProductOperator(_At, &A, true, false);
+      auto _AtA_reg = new SumOperator(_AtA, 1.0, _M.As<Operator>(), shift, true,
+                                      false);
+      AtA_reg.Reset(_AtA_reg, true);
+
+      // Block operator
+      if (it_solver->GetComm() != MPI_COMM_NULL)
+      {
+         block_solver.Reset(new SchurConstrainedSolver(it_solver->GetComm(), *_AtA_reg,
+                                                       *const_cast<DenseMatrix*>(&C), solver), true);
+      }
+      else
+      {
+         block_solver.Reset(new SchurConstrainedSolver(*_AtA_reg,
+                                                       *const_cast<DenseMatrix*>(&C), solver), true);
+      }
+   }
+   else if (dynamic_cast<DenseMatrixInverse*>(&solver))
+   {
+      auto _AtA_reg = new DenseMatrix(A.Width());
+      Vector col_i(A.Height()), col_j(A.Height());
+      for (int i = 0; i < A.Width(); i++)
+      {
+         A.GetColumn(i, col_i);
+         for (int j = 0; j < A.Width(); j++)
+         {
+            A.GetColumn(j, col_j);
+            (*_AtA_reg)(i,j) = col_i * col_j + (M?(*M)(i,j):(i==j))*shift;
+         }
+      }
+      AtA_reg.Reset(_AtA_reg, true);
+
+      block_solver.Reset(new SchurConstrainedSolver(*_AtA_reg,
+                                                    *const_cast<DenseMatrix*>(&C), solver), true);
+   }
+   else { mfem_error("Solver not implemented for this wrapper!"); }
+
+   block_solver.As<SchurConstrainedSolver>()->LagrangeSystemMult(rhs, z);
+
+   x.SetSize(A.Width());
+   y.SetSize(C.Width());
+   x = 0.0;
+   y = 0.0;
+
+   x += z.GetBlock(0);
+   y += z.GetBlock(1);
+}
+
+/// @brief Dense small least squares solver, with constrains @a C with value @a c
+/// but the constraint is one-dimensional, and passed as a vector. Does not return
+/// the multiplier.
+void LSSolver(Solver& solver,
```

Comment:
To disambiguate the `solver` argument from the function which has Solver in the name, I suggest renaming to `linear_solver` (applies to all other use of `solver` as well)